gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/TimeAwareRecursiveCopyableDataset.java

```diff
@@ -0,0 +1,71 @@
+package org.apache.gobblin.data.management.copy;
+
+import com.google.common.collect.Lists;
+import java.io.IOException;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Properties;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.PathFilter;
+
+public class TimeAwareRecursiveCopyableDataset extends RecursiveCopyableDataset {
+  private static final String CONFIG_PREFIX = CopyConfiguration.COPY_PREFIX + ".recursive";
+  public static final String DATE_PATTERN_KEY=CONFIG_PREFIX + ".date.pattern";
+  public static final String LOOKBACK_DAYS_KEY=CONFIG_PREFIX + ".lookback.days";
+
+  private final Integer lookbackDays;
+  private final String datePattern;
+
+  public TimeAwareRecursiveCopyableDataset(FileSystem fs, Path rootPath, Properties properties, Path glob) {
+    super(fs, rootPath, properties, glob);
+    this.lookbackDays = Integer.parseInt(properties.getProperty(LOOKBACK_DAYS_KEY));
+    this.datePattern = properties.getProperty(DATE_PATTERN_KEY);
+  }
+
+  public static class DateRangeIterator implements Iterator {
+    private LocalDateTime startDate;
+    private LocalDateTime endDate;
+    private boolean isDatePatternHourly;
+
+    public DateRangeIterator(LocalDateTime startDate, LocalDateTime endDate, String datePattern) {
+      this.startDate = startDate;
+      this.endDate = endDate;
+      this.isDatePatternHourly = datePattern.contains("HH")? true: false;
```

Comment:
This is not fully correct, consider the pattern:
```
`hourlyHH`/yyy/MM/dd
```
where `hourlyHH` is a string literal in the pattern (we allow string literals, as sometimes there may be prefixes, infixes, or suffixes in the pattern). This will resolve to true, even though the pattern is not actually hourly.

A different way of achieving this is using `DateTimeFormatter#getResolverFields`, then checking whether any of the resolver fields is an hourly field.

Reply:
Thanks for the feedback! There are 2 points to be mentioned here:
1. Any invalid dateTime pattern will result in an IllegalArgumentException to be thrown.
2. To check if input dateTime pattern has hour granularity, DateTimeFormatter#getResolverFields does not help. I created a helper method that passes a reference dateTime (with non zero hour) through the input pattern and then passing the same dateTime (with zeroed hour) through the same pattern and checking if the two output strings are the same. If not, the input dateTime pattern has hourly granularity.